File: app/src/main/java/com/ofalvai/bpinfo/ui/alertlist/AlertListPagerAdapter.java
package com.ofalvai.bpinfo.ui.alertlist;

import android.content.Context;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.util.SparseArray;
import android.view.ViewGroup;

import com.ofalvai.bpinfo.R;

public class AlertListPagerAdapter extends FragmentPagerAdapter implements ViewPager.OnPageChangeListener {

    private static final int FRAGMENT_COUNT = 2;

    private Context mContext;

    // http://stackoverflow.com/a/15261142/745637
    private SparseArray<Fragment> mRegisteredFragments = new SparseArray<>();

    public AlertListPagerAdapter(FragmentManager fm, Context context) {
        super(fm);
        mContext = context;
    }

    @Override
    public Fragment getItem(int position) {
        Fragment fragment;
        switch (position) {
            case 0:
                fragment = AlertListFragment.newInstance(AlertListType.ALERTS_TODAY);
                mRegisteredFragments.put(position, fragment);
                break;
            case 1:
                fragment = AlertListFragment.newInstance(AlertListType.ALERTS_FUTURE);
                mRegisteredFragments.put(position, fragment);
                break;
            default:
                fragment = new Fragment();
        }
        return fragment;
    }

    // http://stackoverflow.com/a/15261142/745637
    @Override
    public Object instantiateItem(ViewGroup container, int position) {
        Fragment fragment = (Fragment) super.instantiateItem(container, position);
        mRegisteredFragments.put(position, fragment);
        return fragment;
    }

    @Override
    public int getCount() {
        return FRAGMENT_COUNT;
    }

    @Override
    public CharSequence getPageTitle(int position) {
        String title;
        switch (position) {
            case 0:
                title = mContext.getString(R.string.tab_title_today);
                break;
            case 1:
                title = mContext.getString(R.string.tab_title_future);
                break;
            default:
                title = "Default tab";
        }
        return title;
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

    }

    @Override
    public void onPageScrollStateChanged(int state) {

    }

    @Override
    public void onPageSelected(int position) {
        AlertListFragment fragment = (AlertListFragment) mRegisteredFragments.get(position);
        fragment.updateSubtitle();

    }
}

File: app/src/main/java/com/ofalvai/bpinfo/ui/alertlist/adapter/AlertListPagerAdapter.kt
/*
 * Copyright 2018 Olivér Falvai
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ofalvai.bpinfo.ui.alertlist.adapter

import android.content.Context
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentPagerAdapter
import android.support.v4.view.ViewPager
import android.util.SparseArray
import android.view.ViewGroup
import com.ofalvai.bpinfo.R
import com.ofalvai.bpinfo.ui.alertlist.AlertListContract
import com.ofalvai.bpinfo.ui.alertlist.AlertListFragment
import com.ofalvai.bpinfo.ui.alertlist.AlertListType

class AlertListPagerAdapter(fm: FragmentManager, private val context: Context)
    : FragmentPagerAdapter(fm), ViewPager.OnPageChangeListener {

    companion object {
        private const val FRAGMENT_COUNT = 2
    }

    // http://stackoverflow.com/a/15261142/745637
    private val registeredFragments = SparseArray<Fragment>()

    override fun getItem(position: Int): Fragment {
        return when (position) {
            0 -> AlertListFragment.newInstance(AlertListType.ALERTS_TODAY)
            1 -> AlertListFragment.newInstance(AlertListType.ALERTS_FUTURE)
            else -> Fragment()
        }
    }

    // http://stackoverflow.com/a/15261142/745637
    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val fragment = super.instantiateItem(container, position) as Fragment
        registeredFragments.put(position, fragment)
        return fragment
    }

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        registeredFragments.remove(position)
        super.destroyItem(container, position, `object`)
    }

    override fun getCount() = FRAGMENT_COUNT

    override fun getPageTitle(position: Int): CharSequence {
        return when (position) {
            0 -> context.getString(R.string.tab_title_today)
            1 -> context.getString(R.string.tab_title_future)
            else -> "Default tab"
        }
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {

    }

    override fun onPageScrollStateChanged(state: Int) {

    }

    override fun onPageSelected(position: Int) {
        val view = registeredFragments.get(position) as? AlertListContract.View
        view?.updateSubtitle()
    }
}
